beacon/sync/src/main/java/tech/pegasys/teku/beacon/sync/historical/ReconstructHistoricalStatesService.java

```diff
@@ -13,18 +13,144 @@
+      final Spec spec,
+      final Optional<String> genesisStateResource) {
+    this.storageUpdateChannel = storageUpdateChannel;
+    this.chainDataClient = chainDataClient;
+    this.spec = spec;
+    this.genesisStateResource = genesisStateResource;
+  }
 
   @Override
   protected SafeFuture<?> doStart() {
-    return SafeFuture.COMPLETE;
+    if (genesisStateResource.isEmpty()) {
+      return SafeFuture.failedFuture(
+          new IllegalStateException("Genesis state resource not provided"));
+    }
+
+    final String resource = genesisStateResource.get();
+    final BeaconState genesisState;
+    try {
+      genesisState = ChainDataLoader.loadState(spec, resource);
+    } catch (IOException e) {
+      LOG.error("Failed to load initial state", e);
+      return SafeFuture.failedFuture(
+          new InvalidConfigurationException(
+              "Failed to load initial state from " + resource + ": " + e.getMessage()));
+    }
+
+    final SafeFuture<Optional<Checkpoint>> future = chainDataClient.getInitialAnchor();
+    future
+        .thenAccept(
+            checkpoint -> {
+              if (checkpoint.isEmpty()) {
+                return; // todo confirm skip
+              }
+
+              final UInt64 anchorSlot = checkpoint.get().getEpochStartSlot(spec);
+              applyBlocks(genesisState, 100, anchorSlot);
+            })
+        .finish(LOG::error); // todo ignores return value
+
+    return SafeFuture.COMPLETE; // todo check
+  }
+
+  public void applyBlocks(
+      final BeaconState genesisState, final int dataStorageFrequency, final UInt64 anchorSlot) {
+    Context context =
+        new Context(genesisState, UInt64.ZERO, UInt64.ONE, dataStorageFrequency, anchorSlot);
+    applyNextBlock(context).finish(LOG::error); // todo ignores return value
+  }
+
+  private SafeFuture<Void> applyNextBlock(Context context) {
+    if (context.slot.isGreaterThanOrEqualTo(context.anchorSlot)) {
+      return SafeFuture.COMPLETE;
+    }
+
+    return chainDataClient
+        .getBlockAtSlotExact(context.slot)
+        .thenAccept(
+            block -> {
+              try {
+                if (block.isEmpty()) {
+                  return;
+                }
+
+                context.currentState =
+                    spec.replayValidatedBlock(context.currentState, block.orElseThrow());
+                if (context.checkStoringHistoricalStates()) {
+                  storageUpdateChannel
+                      .onFinalizedState(context.currentState)
+                      .finish(LOG::error); // todo ignores return value
+                }
+
+                context.incrementSlot();
+
+              } catch (StateTransitionException e) {
+                LOG.error(e); // todo check approach - crash
+              }
+            })
+        .thenCompose(__ -> applyNextBlock(context));
   }
 
   @Override
   protected SafeFuture<?> doStop() {
-    return SafeFuture.COMPLETE;
+    return null;
```

Comment:
For now stick with:
```suggestion
    return SafeFuture.COMPLETE;
```

though we will need to work out how to actually stop properly I suspect.